Feature/get node properties

albertshau opened the pull request.

albertshau committed:
adding endpoint to get node properties from a cluster.

albertshau committed:
remove unused code.

sibonli commented on server/src/main/java/com/continuuity/loom/http/LoomRPCHandler.java:
```diff
@@ -90,4 +101,105 @@ public void getClusterStatuses(HttpRequest request, HttpResponder responder) thr
+
+    JsonObject output = new JsonObject();
+    Set<Node> clusterNodes = clusterService.getClusterNodes(clusterId, userId);
+    if (clusterNodes.isEmpty()) {
```
Is this statement necessary? If it is empty then it should ignore the for each loop and send the same response at the end.

albertshau replied:
good catch, will remove it

sibonli commented on server/src/main/java/com/continuuity/loom/http/LoomRPCHandler.java:
```diff
@@ -90,4 +101,105 @@ public void getClusterStatuses(HttpRequest request, HttpResponder responder) thr
+   */
+  @POST
+  @Path("/getNodeProperties")
+  public void getClusterNodes(HttpRequest request, HttpResponder responder) throws Exception {
```
Would getNodeProperties() be a better name for this method?

albertshau replied:
it would, will change

sibonli commented on server/src/main/java/com/continuuity/loom/store/SQLClusterStore.java:
```diff
@@ -700,6 +700,35 @@ public void deleteNode(String nodeId) throws Exception {
+          while (rs.next()) {
+            Blob blob = rs.getBlob(1);
+            Node node = deserializeBlob(blob, Node.class);
+            nodes.add(node);
```
This result deserialization block is seen in most of the methods in this class. Might be better create a separate method that is reused.

albertshau replied:
agreed, will do

sibonli commented on server/src/test/java/com/continuuity/loom/TestHelper.java:
```diff
@@ -56,4 +59,20 @@ public static void finishTask(String loomUrl, JsonObject jsonObject) throws Exce
+    JsonObject output = new JsonObject();
+    for (int i = 0; i < keyvalues.length; i += 2) {
+      output.addProperty(keyvalues[i], keyvalues[i + 1]);
+    }
```
Would be better if this method took in Map.Entries or some sort of K-V pair rather than a list of strings.

albertshau replied:
this was purely a convenience method for testing so I'd rather not create objects to send in as that would defeat the purpose.  I might just remove this as its used in only one place for an object with 2 keyvalues

sibonli commented:
A few comments

chtyim commented on server/src/main/java/com/continuuity/loom/http/LoomRPCHandler.java:
```diff
@@ -90,4 +101,105 @@ public void getClusterStatuses(HttpRequest request, HttpResponder responder) thr
+    try {
+      body = GSON.fromJson(reader, JsonObject.class);
+    } catch (Exception e) {
+      responder.sendError(HttpResponseStatus.BAD_REQUEST, "invalid request body. Must be a valid JSON Object.");
```
invalid -> Invalid

albertshau replied:
done

chtyim commented on server/src/main/java/com/continuuity/loom/http/LoomRPCHandler.java:
```diff
@@ -90,4 +101,105 @@ public void getClusterStatuses(HttpRequest request, HttpResponder responder) thr
+    Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()), Charsets.UTF_8);
+    JsonObject body;
+    try {
+      body = GSON.fromJson(reader, JsonObject.class);
```
Why deserialize into JsonObject type? Why not use a class type that has the expected structure in it? E.g.

``` java
class ClusterNodesRequest {
  String clusterId;
  Set<String> properties;
  Set<String> services;
}
```

This would simplify a lot of the manual decoding logic here and make the code more reusable in future (say, you could encapsulate encode/decode logic into TypeAdapter if needed).

albertshau replied:
good point, will do

chtyim commented on server/src/main/java/com/continuuity/loom/http/LoomRPCHandler.java:
```diff
@@ -90,4 +101,105 @@ public void getClusterStatuses(HttpRequest request, HttpResponder responder) thr
+
+      // if the node has all services needed
+      if (nodeServices.containsAll(requiredServices)) {
+        JsonObject outputProperties;
```
Try not to use Json type as data structure. E.g. this one is a `Map<String, String>`, isn't it? Using JsonObject everywhere makes the code highly difficult to read.

albertshau replied:
this one isn't a map of strings, the node properties can be populated by plugins further down and the only restriction we've set is that its a json object, so the values here can be any json element.

chtyim commented on server/src/main/java/com/continuuity/loom/http/LoomRPCHandler.java:
```diff
@@ -90,4 +101,105 @@ public void getClusterStatuses(HttpRequest request, HttpResponder responder) thr
+    responder.sendJson(HttpResponseStatus.OK, output);
+  }
+
+  private Set<String> getSet(JsonObject object, String key) {
```
We shouldn't be doing this kind of manual decode. See my comment above.

albertshau replied:
done

chtyim commented on server/src/main/java/com/continuuity/loom/store/ClusterStore.java:
```diff
@@ -189,6 +189,16 @@
+   *         Empty if no cluster or cluster nodes exist.
+   * @throws Exception if there was a problem getting the cluster nodes.
+   */
+  Set<Node> getClusterNodes(String clusterId, String userId) throws Exception;
```
Is the returned Set mutable or immutable or unmodifiable (difference between the last two is immutable means it's will not change, unmodifiable means the content might get change internal, more like a view).

In general when returning Collection type object from a method, it's always important to state what kind of behavior of the instance has.

chtyim replied:
Since this is an interface, it'd be easiest to say returned Set is immutable (as it impose less requirement on the implementation).

albertshau replied:
got it, will do

albertshau committed:
The node properties request is now its own object with a
deserializer to avoid custom decoding. Refactored cluster store
to return immutable collections and consolidated a lot of common
code into their own methods.

albertshau committed:
removing unneeded method and adding class that I forgot to add.

albertshau committed:
style fix.

albertshau committed:
copy immutable set from a set instead of list.

albertshau committed:
remove unintended change.

chtyim commented:
LGTM.

wolf31o2 commented:
Resolves #4
